### ISSUE TITLE
Rule request: complete NSNotificationCenter detachment

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#1061]
From http://fauxpasapp.com/rules/#rule-CompleteNotificationCenterDetachment:

---

Warns when an object removes itself as an observer to all notifications (by invoking `-[NSNotificationCenter removeObserver:]` with `self`).

This can be a problem if a superclass or a subclass wants to keep observing some notifications. The recommended way to handle this is to remove self as an observer only for the specific notifications that you have previously begun observing.

This rule does not warn about such removals occuring in `-dealloc`.

**Comment by mRs-** [comment]
I don't understand this warning exactly. You write 
> This can be a problem if a superclass or a subclass wants to keep observing some notifications. 
IMHO if the dealloc of the super or the subclass is called it's pretty normal to remove the observers here, because the Object (even the sub or the super class) is gone from memory. It's enforcing the Observer to be in the notification center pointing on dead code.

**Comment by marcelofabri** [comment]
Imagine you have a View Controller. In `viewDidDisappear:`, you call `NotificationCenter.default.removeObserver(self)`.

This will unregister the view controller from ALL notifications, including the ones that the superclass registered. So if `UIViewController` actually needs to listen for any notification, it won't get it.

That's way this is an issue. The only place it's safe to do that is in `dealloc`, because you know that the object won't receive any notifications after that anyway.

**Comment by mRs-** [comment]
Yeah but here it's the same principle. If my subclass is calling `viewDidDisappear` it will be called in my superclass as well. So this warning is completely useless.

**Comment by marcelofabri** [comment]
It's not, because `viewDidDisappear` from the superclass will not register for notifications again.

Check the documentation: https://developer.apple.com/documentation/foundation/nsnotificationcenter/1413994-removeobserver?language=objc

**Comment by mRs-** [comment]
Okay now I get it. But the rule is wrong anyway. It's implemented for objective-c code and the right place to do this would be `deinit` and not `dealloc`

**Comment by mRs-** [comment]
I will create a Pull Request to fix this issue.

**Comment by marcelofabri** [comment]
It's not implemented wrong, just the description is using Objective-C in this issue. Check the [examples](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/NotificationCenterDetachmentRuleExamples.swift).

**Comment by mRs-** [comment]
Sorry it was my bad the code looked like this… 
```swift
deinit {
    unregisterForNotification()
}

private func unregisterForNotification() {
  NotificationCenter.default.removeObserver(self) // creates a warning of course
}
```

Thanks for helping out @marcelofabri

**Comment by marcelofabri** [comment]
No problem!

Unfortunately we can't (yet?) recognize this case, so this rule is pretty dumb in general (that's why it's opt-in).

**Comment by kocyigityunus** [comment]
this behavior is absolutely correct. do not remove all observers. just remove the observers you are responsible for with 

open func removeObserver(_ observer: Any, name aName: NSNotification.Name?, object anObject: Any?)

method.

**Comment by WaylanSands** [comment]
I know this is quite old though I'm getting warnings to add my remove observers method to `deinit`. I thought that `deinit` will not be called because the the view controller has observers that are referencing it therefore you can never remove the observes or the view controller from memory?
